src/types/timezone.jl

```diff
@@ -19,10 +25,22 @@ const THREAD_TZ_CACHES = Vector{Dict{String,Tuple{TimeZone,Class}}}()
 end
 @noinline _tz_cache_length_assert() = @assert false "0 < tid <= length(THREAD_TZ_CACHES)"
 
-function _reset_tz_cache()
-    # ensures that we didn't save a bad object
+function _tz_cache_init()
     resize!(empty!(THREAD_TZ_CACHES), Threads.nthreads())
 end
+# ensures that we didn't save a bad object
+function _reset_tz_cache()
+    # Since we use thread-local caches, we spawn a task on _each thread_ to clear that
+    # thread's local cache.
+    Threads.@threads for i in 1:Threads.nthreads()
+        @assert Threads.threadid() === i "TimeZones.TZData.compile() must be called from the main, top-level Task."
+        empty!(_tz_cache())
+    end
+    @lock tz_cache_mutex begin
```

Comment:
Using the function `lock` may avoid the compat issue

Reply:
Yeah, that's a good suggestion. We've been using `Base.@lock` instead internally, to avoid the potential overhead from the closure..

And indeed, we need to mutate a local variable inside the locked body down below:
https://github.com/JuliaTime/TimeZones.jl/blob/6c0d3ead7640f7818b7cbea9dd3596423f73c225/src/types/timezone.jl#L94-L103

So using `Base.lock(l) do ...` would end up allocating on every call. 😕 👎 😞 


So I'd probably opt to keep it how it currently is. Do you have any preference on that? Should I file a PR to add `Base.@lock` to compat?

Ref: https://github.com/JuliaLang/julia/issues/36441 and https://github.com/JuliaLang/julia/pull/39588

Reply:
Lets add `@lock` to Compat.jl so everyone can use this

Reply:
Thanks @omus, I've filed https://github.com/JuliaLang/Compat.jl/issues/762 to track that. I can probably do it myself, but don't want to block this PR since it's already taken me so long to get back to it! 😅 😊

Reply:
Oh, but actually, even better: now that you've lower-bounded to julia 1.6+, i can drop this compat thing entirely. 💪 🎉